FizzBuzz/Program.cs

```diff
@@ -1,12 +1,163 @@
 ﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
 
 namespace FizzBuzz
 {
+    public class FizzBuzz : IEnumerable<string>
+    {
+        private List<string> items;
+        private Dictionary<string, bool> rules;
+
+        public FizzBuzz(int n, Dictionary<string, bool> _rules)
+        {
+            items = new List<string>();
+            rules = _rules;
+
+            for (int i = 1; i <= n; ++i)
+            {
+                items.Add(GetFizzBuzz(i));
+            }
+        }
+
+        public string this[int index]
+        {
+            get { return items[index]; }
+            set { items.Insert(index, GetFizzBuzz(index)); }
+        }
+
+        public IEnumerator<string> GetEnumerator()
+        {
+            return items.GetEnumerator();
+        }
+
+        IEnumerator IEnumerable.GetEnumerator()
+        {
+            return this.GetEnumerator();
+        }
+
+        private string GetFizzBuzz(int n)

```

Comment:
One way of making this prettier and more OOP-y, would be to:
* Create a `Rule` interface/base class, with a method `Apply`
* Make classes for each rule
* This would essentially be `foreach rule in rules { ans = rule.apply(ans) }`
* You will need to be careful about the order in which you apply them, but I think it should be doable 
* If `Rule` is a class, you can add custom rules for "if number is multiple of X, print Y", and implement them there